src/utils/fs.js

```diff
@@ -2,38 +2,82 @@ import Promise from 'es6-promise/lib/es6-promise/promise.js';
 import * as fs from 'fs';
 import { dirname } from './path.js';
 
+function readdir ( dir ) {
+	return new Promise( ( fulfil, reject ) => {
+		fs.readdir( dir, ( err ) => {
+			if ( err ) {
+				reject( err );
+			} else {
+				fulfil();
+			}
+		});
+	});
+}
+
+function mkdir ( dir ) {
+	return new Promise( ( fulfil, reject ) => {
+		fs.mkdir( dir, ( err ) => {
+			if ( err ) {
+				reject( err );
+			} else {
+				fulfil();
+			}
+		});
+	});
+}
+
 function mkdirpath ( path ) {
 	const dir = dirname( path );
-	try {
-		fs.readdirSync( dir );
-	} catch ( err ) {
-		mkdirpath( dir );
-		fs.mkdirSync( dir );
-	}
+	return readdir( dir ).catch( ()=>{
+		return mkdirpath( dir ).then( ()=>{
+			return mkdir( dir ).catch( e=>{
+				// race condition
+				if (e && e.code === 'EEXIST') {
+					return;
+				}
+				throw e;
+			});
+		});
+	});
 }
 
 export function isFile ( file ) {
-	try {
-		const stats = fs.statSync( file );
-		return stats.isFile();
-	} catch ( err ) {
-		return false;
-	}
+	return new Promise( ( fulfil, reject ) => {
+		fs.stat( file, ( err, stats ) => {
+			if ( err ) {
+				reject( err );
```

Comment:
This API doesn't feel correct to me. I typically write async functions using the same API as the synchronous ones, but with async semantics. If this were synchronous (i.e. what's on master now), it would return `true` or `false` to answer the question and throw an exception in some other weird case. It would not return the filename on success and throw otherwise. I see that this is just to make `addJsExtensionIfNecessary`'s job easier, but I don't think that's a valid reason to have this API.